src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1413 @@
+                foreach (var baseType in typename.TypeDefinitionAst.BaseTypes)
+                {
+                    var baseTypeName = baseType.TypeName as TypeName;
+                    if (baseTypeName == null) continue;
+                    var baseTypeDefinitionAst = baseTypeName._typeDefinitionAst;
+                    results.AddRange(GetMembersByInferredType(new PSTypeName(baseTypeDefinitionAst), @static, filterToCall));
+                }
+
+                // Add stuff from our base class System.Object.
+                if (@static)
+                {
+                    // Don't add base class constructors
+                    if (filter == null)
+                    {
+                        filterToCall = o => !IsConstructor(o);
+                    }
+                    else
+                    {
+                        filterToCall = o => !IsConstructor(o) && filter(o);
+                    }
+
+                    if (!foundConstructor)
+                    {
+                        results.Add(
+                            new CompilerGeneratedMemberFunctionAst(PositionUtilities.EmptyExtent, typename.TypeDefinitionAst,
+                                SpecialMemberFunctionType.DefaultConstructor));
+                    }
+                }
+                else
+                {
+                    // Reset the filter because the recursive call will add IsHidden back if necessary.
+                    filterToCall = filter;
+                }
+                results.AddRange(GetMembersByInferredType(new PSTypeName(typeof(object)), @static, filterToCall));
+            }
+            else
+            {
+                // Look in the type table first.
+                if (!@static)
+                {
+                    var consolidatedString = new ConsolidatedString(new[] { typename.Name });
+                    results.AddRange(ExecutionContext.TypeTable.GetMembers<PSMemberInfo>(consolidatedString));
+                }
+
+                string cimNamespace;
+                string className;
+                if (ParseCimCommandsTypeName(typename, out cimNamespace, out className))
+                {
+                    AddCommandWithPreferenceSetting(_powerShell, "CimCmdlets\\Get-CimClass")
+                        .AddParameter("Namespace", cimNamespace)
+                        .AddParameter("Class", className);
+                    Exception unused;
+                    var classes = Helper.ExecuteCurrentPowerShell(out unused);
+                    foreach (var @class in classes.Select(PSObject.Base).OfType<CimClass>())
+                    {
+                        results.AddRange(filterToCall != null ? @class.CimClassProperties.Where(filterToCall) : @class.CimClassProperties);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        internal static PowerShell AddCommandWithPreferenceSetting(PowerShell powershell, string command, Type type = null)
```

Comment:
This method copy-pasted from `src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs`, why?